Archive/Draft/RFCxxxx-PS7-Bootstrapper.md

```diff
@@ -0,0 +1,256 @@
+---
+RFC: RFCnnnn
+Author: Steve Lee
+Status: Draft
+SupercededBy: n/a
+Version: 0.1
+Area: Install
+Comments Due: Dec 31st, 2021
+Plan to implement: Yes
+---
+
+# Bootstrapper for PS7 in Windows
+
+## Motivation
+
+    As an admin,
+    I can use PS7 to provision a new Windows machine,
+    so that I can leverage new capabilities in PS7.
+
+    As a user,
+    I can use PS7 on a new Windows machine,
+    so that I can leverage new capabilities in PS7.
+
+Explicitly installing PS7 is a blocker for some customers not only due to discoverability, but also some customers have policies that
+don't allow for installing software onto Windows after initial provisioning.
+
+## Shipping in Windows
+
+There are several issues that prevent simply shipping PS7 in Windows:
+
+- Size
+- Support lifecycle differences
+
+### Size constraints
+
+A complete installation of PowerShell 7 currently is around 200 MB (unpacked).
+This is due to including everything (modules and .NET assemblies) to provide parity with what Windows PowerShell 5.1 with .NET Framework provides.
+There is currently no plan by Windows or the .NET team to ship newer .NET in Windows, so PS7 will need to continue to ship .NET with PS7.
+
+### Support lifecycle
+
+Windows has a 5+5 year support lifecycle.
+.NET supports up to 3 years for Long Term Support (LTS) versions.
+So even if size was not an issue, we cannot check in a version of PS7 with a .NET that is not supported while Windows itself is supported.
+
+### Solution
+
+To solve both of these issues, the proposal is to ship a bootstrapper in Windows that would be supported within the Windows support lifecycle.
+The boostrapper will update PS7 as needed to ensure a supported version of .NET is being used.
+Details of this bootstrapper follow.
+
+## Specification
+
+### pwsh.exe bootstrapper
+
+We would check-in a `pwsh.exe` into Windows 11+ (under `System32`) written in [native code](#native-code) that will:
+
+- Detect if PS7 is already installed
+  - If installed, pass-through all arguments to the new detached `pwsh` child process
+    - This means that if PS7 is already in the `PATH`, then that version (whether installed via MSI, MSIX, or other means)
+      needs to be updated outside of the bootstrapper
+  - Otherwise, [install](#installer) PS7 and then pass-through all arguments to the new detached `pwsh` child process
+
+The boostrapper should explicitly create a detached child `pwsh` process and then exit.
+As a detached child process, it will continue execution even after the parent has exited.
+
+Expectation is that the bootstrapper will always be first in the `$env:PATH` and executed first, so it can also serve
```

Comment:
Won't this add even more time to executing `pwsh`, you now have to wait for 2 processes to initialise and while it's minute it's still a problem today with just `pwsh.exe` itself.

Reply:
It will add some time, but should be minimal impact as the shim is in native code, but there is a cost to starting a 2nd process, but I think overall it'll be minimal compared to the current limitations of pwsh startup with .NET

Reply:
I think @awakecoding's comment has a good point. He suggests that, if we decide to go with the shim approach, then the bootstrapper `pwsh` can actually host CoreCLR runtime, and calls into `pwsh.dll` via its entry point. In this way, we don't need to spin up 2 processes and handle STD io redirections.

Reply:
@daxian-dbw it's actually even better than that: you can literally use the hostfxr APIs to call into the pwsh.exe command-line entry point... within the current process. You literally load pwsh.exe and launch it within the PowerShell native host process without a subprocess! When using this specific mode of launching with a PowerShell native host, you can only do it as a one-shot thing where you basically just call it once and let it return, but that's precisely what we want here.

Reply:
Thank @awakecoding, I just saw the `run_pwsh_lib` and `run_pwsh_app` in your project. What's the benefit of loading and launching `pwsh.exe` comparing to loading `Microsoft.PowerShell.ConsoleHost.dll` and calling into the entry point function [`Start(string[] args, int argc)`](https://github.com/PowerShell/PowerShell/blob/c2966acfd570a37076c28746db35a3bdcbd15412/src/Microsoft.PowerShell.ConsoleHost/host/msh/ManagedEntrance.cs#L50)? The former will load extra libraries (`pwsh.exe` and `pwsh.dll`).

Reply:
Storing the bootstrapper version of pwsh that early in the $PATH variable seems odd. Wouldn't it make more sense to put it closer to the end? This way the real pwsh.exe can be executed instead of the bootstrapper when both are installed?

Reply:
> Thank @awakecoding, I just saw the `run_pwsh_lib` and `run_pwsh_app` in your project. What's the benefit of loading and launching `pwsh.exe` comparing to loading `Microsoft.PowerShell.ConsoleHost.dll` and calling into the entry point function [`Start(string[] args, int argc)`](https://github.com/PowerShell/PowerShell/blob/c2966acfd570a37076c28746db35a3bdcbd15412/src/Microsoft.PowerShell.ConsoleHost/host/msh/ManagedEntrance.cs#L50)? The former will load extra libraries (`pwsh.exe` and `pwsh.dll`).

There is currently no way to load less than what pwsh.exe needs, and even if there was a way to do it (lazy loading of DLLs, etc) then the proper way would still be with a .NET native host process, so we're back to my technique again.

Also, how would you load ConsoleHost.dll from a process that isn't .NET, and without creating a subprocess? Chances are you would need to use the hostfxr APIs again to do this cleanly, and then you'd be limited to the command-line interface.

The PowerShell native host approach is the only one that avoids creating a subprocess and solves the problem very efficiently. Think of it as just replacing pwsh.exe out of the full PowerShell installation, except it doesn't rely on .NET at all, making it the perfect shim.

Reply:
> Also, how would you load ConsoleHost.dll from a process that isn't .NET, and without creating a subprocess? Chances are you would need to use the hostfxr APIs again to do this cleanly, and then you'd be limited to the command-line interface.

The `pwrshplugin` used by PowerShell for WSMan remoting loads `System.Management.Automation.dll` directly from native process to create a function pointer, see code [here](https://github.com/PowerShell/PowerShell-Native/blob/master/src/powershell-native/nativemsh/pwrshplugin/pwrshclrhost.cpp#L58-L62).

We also have an ancient native powershell host written in C++, and it used to load `ConsoleHost.dll` directly from native process, though it hasn't been used for a long time. See the code [here](https://github.com/PowerShell/PowerShell-Native/blob/1d57ad0f5dea86cc5154aadf68932230f5bb6ab1/src/powershell-native/nativemsh/pwrshexe/CssMainEntry.cpp#L56-L61).

We also have the DSC native agent that loads CoreCLR and calls into managed code form native code. So, I think it's doable, only that we need to figure out how to get the Trusted Assembly List (TPA) populated by using `hostfx`.

Reply:
This is a similar approach that shouldn't change much, the sample you linked to appears to be using COM Interop to load PowerShell. While I believe it is still possible to use COM Interop on Windows with PowerShell 7, I do recall trying it on Linux only to realize the COM exports were missing on non-Windows. 

The current project is only for Windows, but it wouldn't hurt to use the recommended hosting APIs, and that would mean the hostfxr APIs. Maybe the only thing we can conclude at this point is to use the hostfxr APIs to make a .NET host, and then move on to *how* we'd load and call PowerShell using it.

Reply:
Agreed. Regardless of the implementation details, I believe we should make the bootstrapper a .NET host, if we want the bootstrapper a shim that always runs.

Reply:
I think having the pwsh bootstrapper be a .NET host is interesting.  If it's simply hosting the consolehost and passes all arguments to it to evaluate, then I would agree that is a fine solution.